File: py-radiate/radiate/selector.py
from typing import Dict
from .component import ComponentBase
from .genome.gene import GeneType

class SelectorBase(ComponentBase):
    def __init__(self, component: str, args: Dict[str, str] = {}, allowed_genes: set[str] = {}):
        super().__init__(component=component, args=args)
        self.allowed_genes = allowed_genes if allowed_genes else GeneType.ALL

    def __str__(self):
        """
        Return a string representation of the selector.
        :return: String representation of the selector.
        """
        return f"Selector(name={self.component}, args={self.args})"

    def __repr__(self):
        """
        Return a detailed string representation of the selector.
        :return: Detailed string representation of the selector.
        """
        return f"SelectorBase(selector={self.component}, args={self.args}, allowed_genes={self.allowed_genes})"

    def __eq__(self, value):
        if not isinstance(value, SelectorBase):
            return False
        return self.component == value.component and self.args == value.args and self.allowed_genes == value.allowed_genes


class TournamentSelector(SelectorBase):
    def __init__(self, k: int = 3):
        """
        Initialize the tournament selector with tournament size.
        :param k: Tournament size.
        """
        super().__init__(component="TournamentSelector", args={"k": k})


class RouletteSelector(SelectorBase):
    def __init__(self):
        """
        Initialize the roulette selector.
        """
        super().__init__(component="RouletteSelector")



class RankSelector(SelectorBase):
    def __init__(self):
        """
        Initialize the rank selector.
        """
        super().__init__(component="RankSelector")



class EliteSelector(SelectorBase):
    def __init__(self):
        """
        Initialize the elite selector.
        """
        super().__init__(component="EliteSelector")



class BoltzmannSelector(SelectorBase):
    def __init__(self, temp: float = 1.0):
        """
        Initialize the Boltzmann selector with temperature.
        :param temp: Temperature for the Boltzmann selector.
        """
        super().__init__(component="BoltzmannSelector", args={"temp": temp})



class StochasticSamplingSelector(SelectorBase):
    def __init__(self):
        """
        Initialize the stochastic sampling selector.
        """
        super().__init__(component="StochasticSamplingSelector")



class LinearRankSelector(SelectorBase):
    def __init__(self, pressure: float = 0.5):
        """
        Initialize the linear rank selector.
        :param pressure: Pressure for the linear rank selector.
        """
        super().__init__(component="LinearRankSelector", args={"pressure": pressure})



class NSGA2Selector(SelectorBase):
    def __init__(self):
        """
        Initialize the NSGA2 selector.
        """
        super().__init__(component="NSGA2Selector")



class TournamentNSGA2Selector(SelectorBase):
    def __init__(self, k: int = 3):
        """
        Initialize the Tournament NSGA2 selector with tournament size.
        :param k: Tournament size.
        """
        super().__init__(component="TournamentNSGA2Selector", args={"k": k})



class SteadyStateSelector(SelectorBase):
    def __init__(self, replacement_count: int = 10):
        """
        Initialize the steady state selector.
        """
        super().__init__(component="SteadyStateSelector", args={"replacement_count": replacement_count})




# from radiate.radiate import PySelector


# class SelectorBase:
#     def __init__(self, selector: PySelector):
#         """
#         Initialize the selector base with a Selector instance.
#         :param selector: An instance of Selector.
#         """
#         self.selector = selector

#     def __str__(self):
#         """
#         Return a string representation of the selector.
#         :return: String representation of the selector.
#         """
#         return f"Selector(name={self.selector.name}, args={self.selector.args})"

#     def __repr__(self):
#         """
#         Return a detailed string representation of the selector.
#         :return: Detailed string representation of the selector.
#         """
#         return f"SelectorBase(selector={self.selector})"

#     def __eq__(self, value):
#         if not isinstance(value, SelectorBase):
#             return False
#         return self.selector == value.selector


# class TournamentSelector(SelectorBase):
#     def __init__(self, k: int = 3):
#         """
#         Initialize the tournament selector with tournament size.
#         :param k: Tournament size.
#         """
#         selector = PySelector.tournament_selector(tournament_size=k)
#         super().__init__(selector)


# class RouletteSelector(SelectorBase):
#     def __init__(self):
#         """
#         Initialize the roulette selector.
#         """
#         selector = PySelector.roulette_wheel_selector()
#         super().__init__(selector)


# class RankSelector(SelectorBase):
#     def __init__(self):
#         """
#         Initialize the rank selector.
#         """
#         selector = PySelector.rank_selector()
#         super().__init__(selector)


# class EliteSelector(SelectorBase):
#     def __init__(self):
#         """
#         Initialize the elite selector.
#         """
#         selector = PySelector.elite_selector()
#         super().__init__(selector)


# class BoltzmannSelector(SelectorBase):
#     def __init__(self, temp: float = 1.0):
#         """
#         Initialize the Boltzmann selector with temperature.
#         :param temp: Temperature for the Boltzmann selector.
#         """
#         selector = PySelector.boltzmann_selector(temp=temp)
#         super().__init__(selector)


# class StochasticSamplingSelector(SelectorBase):
#     def __init__(self):
#         """
#         Initialize the stochastic sampling selector.
#         """
#         selector = PySelector.stochastic_sampling_selector()
#         super().__init__(selector)


# class LinearRankSelector(SelectorBase):
#     def __init__(self, pressure: float = 0.5):
#         """
#         Initialize the linear rank selector.
#         :param pressure: Pressure for the linear rank selector.
#         """
#         selector = PySelector.linear_rank_selector(pressure=pressure)
#         super().__init__(selector)


# class NSGA2Selector(SelectorBase):
#     def __init__(self):
#         """
#         Initialize the NSGA2 selector.
#         """
#         selector = PySelector.nsga2_selector()
#         super().__init__(selector)


# class TournamentNSGA2Selector(SelectorBase):
#     def __init__(self, k: int = 3):
#         """
#         Initialize the Tournament NSGA2 selector with tournament size.
#         :param k: Tournament size.
#         """
#         selector = PySelector.tournament_nsga2_selector(tournament_size=k)
#         super().__init__(selector)


# class SteadyStateSelector(SelectorBase):
#     def __init__(self, replacement_count: int = 10):
#         """
#         Initialize the steady state selector.
#         """
#         selector = PySelector.steady_state_selector(replacement_count=replacement_count)
#         super().__init__(selector)

File: crates/radiate-python/src/builder/selector.rs
use crate::PyEngineParam;
use radiate::{
    BoltzmannSelector, Chromosome, EliteSelector, Epoch, GeneticEngineBuilder, LinearRankSelector,
    NSGA2Selector, RankSelector, RouletteSelector, StochasticUniversalSamplingSelector,
    TournamentSelector,
};

use super::ParamMapper;

const TOURNAMENT_SELECTOR: &str = "tournament";
const ROULETTE_SELECTOR: &str = "roulette";
const RANK_SELECTOR: &str = "rank";
const ELITISM_SELECTOR: &str = "elitism";
const BOLTZMANN_SELECTOR: &str = "boltzmann";
const STOCHASTIC_UNIVERSAL_SAMPLING_SELECTOR: &str = "stochastic_universal_sampling";
const LINEAR_RANK_SELECTOR: &str = "linear_rank";
const NSGA2_SELECTOR: &str = "nsga2";

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum SelectorType {
    Tournament,
    Roulette,
    Rank,
    Elitism,
    Boltzmann,
    StochasticUniversalSampling,
    LinearRank,
    NSGA2,
}

pub enum SelectorConfig {
    Tournament(TournamentSelectorMapper),
    Roulette(RouletteSelectorMapper),
    Rank(RankSelectorMapper),
    Elitism(EliteSelectorMapper),
    Boltzmann(BoltzmannSelectorMapper),
    StochasticUniversalSampling(StochasticUniversalSamplingSelectorMapper),
    LinearRank(LinearRankSelectorMapper),
    NSGA2(NSGA2SelectorMapper),
}

pub(super) fn selector_name_to_type(name: &str) -> SelectorType {
    match name {
        TOURNAMENT_SELECTOR => SelectorType::Tournament,
        ROULETTE_SELECTOR => SelectorType::Roulette,
        RANK_SELECTOR => SelectorType::Rank,
        ELITISM_SELECTOR => SelectorType::Elitism,
        BOLTZMANN_SELECTOR => SelectorType::Boltzmann,
        STOCHASTIC_UNIVERSAL_SAMPLING_SELECTOR => SelectorType::StochasticUniversalSampling,
        LINEAR_RANK_SELECTOR => SelectorType::LinearRank,
        NSGA2_SELECTOR => SelectorType::NSGA2,
        _ => panic!("Unknown selector type"),
    }
}

pub struct TournamentSelectorMapper;

impl<C: Chromosome> ParamMapper<C> for TournamentSelectorMapper {
    type Output = TournamentSelector;

    fn map(&self, param: &PyEngineParam) -> Self::Output {
        let args = param.get_args();
        let tournament_size = args
            .get("k")
            .and_then(|s| s.parse::<usize>().ok())
            .unwrap_or(2);
        TournamentSelector::new(tournament_size)
    }
}

pub struct RouletteSelectorMapper;
impl<C: Chromosome> ParamMapper<C> for RouletteSelectorMapper {
    type Output = RouletteSelector;

    fn map(&self, _: &PyEngineParam) -> Self::Output {
        RouletteSelector::new()
    }
}

pub struct RankSelectorMapper;
impl<C: Chromosome> ParamMapper<C> for RankSelectorMapper {
    type Output = RankSelector;

    fn map(&self, _: &PyEngineParam) -> Self::Output {
        RankSelector::new()
    }
}

pub struct EliteSelectorMapper;
impl<C: Chromosome> ParamMapper<C> for EliteSelectorMapper {
    type Output = EliteSelector;

    fn map(&self, _: &PyEngineParam) -> Self::Output {
        EliteSelector::new()
    }
}

pub struct BoltzmannSelectorMapper;
impl<C: Chromosome> ParamMapper<C> for BoltzmannSelectorMapper {
    type Output = BoltzmannSelector;

    fn map(&self, param: &PyEngineParam) -> Self::Output {
        let args = param.get_args();
        let temperature = args
            .get("temp")
            .and_then(|s| s.parse::<f32>().ok())
            .unwrap_or(1.0);
        BoltzmannSelector::new(temperature)
    }
}

pub struct StochasticUniversalSamplingSelectorMapper;
impl<C: Chromosome> ParamMapper<C> for StochasticUniversalSamplingSelectorMapper {
    type Output = StochasticUniversalSamplingSelector;

    fn map(&self, _: &PyEngineParam) -> Self::Output {
        StochasticUniversalSamplingSelector::new()
    }
}

pub struct LinearRankSelectorMapper;
impl<C: Chromosome> ParamMapper<C> for LinearRankSelectorMapper {
    type Output = LinearRankSelector;

    fn map(&self, param: &PyEngineParam) -> Self::Output {
        let args = param.get_args();
        let selection_pressure = args
            .get("pressure")
            .and_then(|s| s.parse::<f32>().ok())
            .unwrap_or(1.0);
        LinearRankSelector::new(selection_pressure)
    }
}

pub struct NSGA2SelectorMapper;
impl<C: Chromosome> ParamMapper<C> for NSGA2SelectorMapper {
    type Output = NSGA2Selector;

    fn map(&self, _: &PyEngineParam) -> Self::Output {
        NSGA2Selector::new()
    }
}

pub(crate) fn set_selector<C, T, E>(
    builder: GeneticEngineBuilder<C, T, E>,
    selector: &PyEngineParam,
    is_offspring: bool,
) -> GeneticEngineBuilder<C, T, E>
where
    C: Chromosome + PartialEq + Clone,
    T: Clone + Send + Sync,
    E: Epoch<Chromosome = C>,
{
    if selector.name() == TOURNAMENT_SELECTOR {
        let args = selector.get_args();
        let tournament_size = args
            .get("k")
            .and_then(|s| s.parse::<usize>().ok())
            .unwrap_or(2);
        return match is_offspring {
            true => builder.offspring_selector(TournamentSelector::new(tournament_size)),
            false => builder.survivor_selector(TournamentSelector::new(tournament_size)),
        };
    } else if selector.name() == ROULETTE_SELECTOR {
        return match is_offspring {
            true => builder.offspring_selector(RouletteSelector::new()),
            false => builder.survivor_selector(RouletteSelector::new()),
        };
    } else if selector.name() == RANK_SELECTOR {
        return match is_offspring {
            true => builder.offspring_selector(RankSelector::new()),
            false => builder.survivor_selector(RankSelector::new()),
        };
    } else if selector.name() == ELITISM_SELECTOR {
        return match is_offspring {
            true => builder.offspring_selector(EliteSelector::new()),
            false => builder.survivor_selector(EliteSelector::new()),
        };
    } else if selector.name() == BOLTZMANN_SELECTOR {
        let args = selector.get_args();
        let temperature = args
            .get("temp")
            .and_then(|s| s.parse::<f32>().ok())
            .unwrap_or(1.0);
        return match is_offspring {
            true => builder.offspring_selector(BoltzmannSelector::new(temperature)),
            false => builder.survivor_selector(BoltzmannSelector::new(temperature)),
        };
    } else if selector.name() == STOCHASTIC_UNIVERSAL_SAMPLING_SELECTOR {
        return match is_offspring {
            true => builder.offspring_selector(StochasticUniversalSamplingSelector::new()),
            false => builder.survivor_selector(StochasticUniversalSamplingSelector::new()),
        };
    } else if selector.name() == LINEAR_RANK_SELECTOR {
        let args = selector.get_args();
        let selection_pressure = args
            .get("pressure")
            .and_then(|s| s.parse::<f32>().ok())
            .unwrap_or(1.0);
        return match is_offspring {
            true => builder.offspring_selector(LinearRankSelector::new(selection_pressure)),
            false => builder.survivor_selector(LinearRankSelector::new(selection_pressure)),
        };
    } else if selector.name() == NSGA2_SELECTOR {
        return match is_offspring {
            true => builder.offspring_selector(NSGA2Selector::new()),
            false => builder.survivor_selector(NSGA2Selector::new()),
        };
    }

    builder
}
